File: python/algorithm/TSP_DP.py
import random


TEST_CASES = 10
N = 10
INF = 987654321


random.seed(43);

cache = [[-1 for _ in range(0,(1<<N))] for _ in range(0,N)]


def initialize() : 

	ret = [ [0 for j in range(0,N)] for i in range(0,N) ];
	ret[1][2] = 3;
	for i in range(0, N) :
		ret[i][i] = INF;
		for j in range(0, i) :
			tmp = random.randint(1, 100)
			if(tmp % 37 < 1) : tmp = INF
			ret[i][j] = tmp;
			ret[j][i] = tmp;



	return ret


def solve(cur, visited, W) :


	if (visited == (1<<N)-1) : return W[cur][0]

	ret = cache[cur][visited];
	if ret != -1 : return ret

	ret = INF

	for i in range(0, N) :
		if not (visited & (1 << i) ) and  W[cur][i] != INF :
			newvisited =  visited + (1<<i)
			ret = min(ret, W[cur][i] + solve(i,  newvisited, W))

	cache[cur][visited] = ret;


	return ret

def print_solution(W, ret) :
	for i in range(0, N):
		tmp = list(map(
			lambda a :   " _ " if a == INF  else "%3d"%a, W[i])
		)
		print(" ".join(tmp))
	print("ret : {}".format(ret))
	print("\n\n")



if __name__ == "__main__"  : 

	for _ in range(0,1) :

		cache = [[-1 for _ in range(0,(1<<N))] for _ in range(0,N)]

		W = initialize()
		visited = 1 
		ret = solve(0, visited, W);
		print_solution(W,ret);

File: c++/algorithm/TSP_DP.cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <limits>
#include <random>
#include <vector>
#define N 10
#define MAX 987654321

using namespace std;

double cache[N][1 << N];

double** initialize() {
  default_random_engine gen(43);
  normal_distribution<double> dis(N, 3 * N);

  double** w = new double*[N];
  for (int i = 0; i < N; i++) {
    w[i] = new double[N];
  }

  for (int i = 0; i < N; i++) {
    w[i][i] = MAX;
    for (int j = 0; j < i; j++) {
      double tmp = abs(dis(gen));
      if ((int)floor(tmp) % 37 < 1) tmp = MAX;

      w[i][j] = tmp;
      w[j][i] = tmp;
    }
  }

  memset(cache, -1, sizeof(cache));

  return w;
}

double shortest_path(int here, int visited, double** w) {
  if (visited == (1 << N) - 1) return w[here][0];

  double& ret = cache[here][visited];

  if (ret >= 0) return ret;

  ret = MAX;

  for (int i = 0; i < N; i++) {
    if (visited & (1 << i)) continue;
    ret = min(ret, shortest_path(i, visited + (1 << i), w) + w[here][i]);
  }

  return ret;
}

void print_result(double** w, double ret) {
  for (int i = 0; i < N; i++) {
    for (int j = 0; j < N; j++) {
      if (w[i][j] == MAX) {
        cout << setw(5) << "-";
        cout << " ";
        continue;
      }
      cout << fixed << setw(5) << setprecision(2) << w[i][j];
      cout << " ";
    }
    cout << endl;
  }

  cout << "result :  " << fixed << setprecision(2) << ret << endl;
}

int main(void) {
  vector<int> path;
  double** w = initialize();
  int visited = 1;
  path.push_back(0);

  double ret = shortest_path(0, visited, w);

  print_result(w, ret);

  return 0;
}